subsys/net/ip/tcp2.c

```diff
@@ -465,113 +466,122 @@ static size_t tcp_data_get(struct tcp *conn, struct net_pkt *pkt)
 	return len;
 }
 
-static void tcp_header_add(struct tcp *conn, struct net_pkt *pkt, u8_t flags)
+static int tcp_finalize_pkt(struct net_pkt *pkt)
 {
-	struct net_buf *buf = net_pkt_get_frag(pkt, K_NO_WAIT);
-	struct tcphdr th;
-
-	memset(&th, 0, sizeof(th));
-
-	th.th_sport = conn->src->sin.sin_port;
-	th.th_dport = conn->dst->sin.sin_port;
-
-	th.th_off = 5;
-	th.th_flags = flags;
-	th.th_win = htons(conn->win);
-	th.th_seq = htonl(conn->seq);
+	net_pkt_cursor_init(pkt);
 
-	if (ACK & flags) {
-		th.th_ack = htonl(conn->ack);
+	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
+		return net_ipv4_finalize(pkt, IPPROTO_TCP);
 	}
 
-	memcpy(net_buf_add(buf, sizeof(th)), &th, sizeof(th));
+	if (IS_ENABLED(CONFIG_NET_IPV6) && net_pkt_family(pkt) == AF_INET6) {
+		return net_ipv6_finalize(pkt, IPPROTO_TCP);
+	}
 
-	net_pkt_frag_insert(pkt, buf);
+	return -EINVAL;
 }
 
-static void ip_header_add(struct tcp *conn, struct net_pkt *pkt)
+static int tcp_header_add(struct tcp *conn, struct net_pkt *pkt, u8_t flags)
 {
-	struct net_buf *buf = net_pkt_get_frag(pkt, K_NO_WAIT);
-
-	pkt->family = net_context_get_family(conn->context);
-
-	switch (pkt->family) {
-	case AF_INET: {
-		struct net_ipv4_hdr ip;
-
-		memset(&ip, 0, sizeof(ip));
-
-		ip.vhl = 0x45;
-		ip.ttl = 64;
-		ip.proto = IPPROTO_TCP;
-		ip.len = htons(net_pkt_get_len(pkt) + sizeof(ip));
-
-		ip.src = conn->src->sin.sin_addr;
-		ip.dst = conn->dst->sin.sin_addr;
+	NET_PKT_DATA_ACCESS_DEFINE(tcp_access, struct tcphdr);
+	struct tcphdr *th;
 
-		memcpy(net_buf_add(buf, sizeof(ip)), &ip, sizeof(ip));
+	th = (struct tcphdr *)net_pkt_get_data(pkt, &tcp_access);
+	if (!th) {
+		return -ENOBUFS;
+	}
 
-		net_pkt_frag_insert(pkt, buf);
+	memset(th, 0, sizeof(struct tcphdr));
 
-		net_pkt_set_ip_hdr_len(pkt, buf->len);
+	th->th_sport = conn->src->sin.sin_port;
+	th->th_dport = conn->dst->sin.sin_port;
 
-		net_pkt_cursor_init(pkt);
+	th->th_off = 5;
+	th->th_flags = flags;
+	th->th_win = htons(conn->win);
+	th->th_seq = htonl(conn->seq);
 
-		net_ipv4_finalize(pkt, IPPROTO_TCP);
-		break;
+	if (ACK & flags) {
+		th->th_ack = htonl(conn->ack);
 	}
-	case AF_INET6: {
-		struct net_ipv6_hdr ip6;
 
-		memset(&ip6, 0, sizeof(ip6));
+	return net_pkt_set_data(pkt, &tcp_access);
+}
 
-		ip6.vtc = 0x60;
-		ip6.nexthdr = IPPROTO_TCP;
-		ip6.len = htons(net_pkt_get_len(pkt));
+static int ip_header_add(struct tcp *conn, struct net_pkt *pkt)
```

Comment:
hopefully this will go away once tcp2 will be properly integrated with net_context

Reply:
yep.